File: problem set 2/readability.c
#include <ctype.h>
#include <cs50.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
void calculate_reading_level(int letters, int words, int sentences);

int main(void)
{
  string text = get_string("Text: ");

  int letters = count_letters(text);
  int words = count_words(text);
  int sentences = count_sentences(text);

  calculate_reading_level(letters, words, sentences);
}

int count_letters(string text)
{
  int letter_count = 0;
  for (int i = 0, len = strlen(text); i < len; i++)
  {
    if (isalpha(text[i]))
    {
      letter_count++;
    }
  }
  return letter_count;
}

int count_words(string text)
{
  int word_count = 1;
  for (int i = 0, len = strlen(text); i < len; i++)
  {
    if (isblank(text[i]))
    {
      word_count++;
    }
  }
  return word_count;
}

int count_sentences(string text)
{
  int sentence_count = 0;
  for (int i = 0, len = strlen(text); i < len; i++)
  {
    if (ispunct(text[i]) && (text[i] == '.' || text[i] == '!' || text[i] == '?'))
    {
      sentence_count++;
    }
  }
  return sentence_count;
}

void calculate_reading_level(int letters, int words, int sentences)
{
  const float L = ((float)letters / (float)words) * 100;
  const float S = ((float)sentences / (float)words) * 100;
  int index = round((0.0588 * L) - (0.296 * S) - 15.8);

  if (index > 16)
  {
    printf("Grade 16+\n");
  }
  else if (index < 1)
  {
    printf("Before Grade 1\n");
  }
  else
  {
    printf("Grade %i\n", index);
  }
}

File: problem set 6/readability.py
from cs50 import get_string


def main():
    text = get_string("Text: ")

    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)

    calculate_reading_level(letters, words, sentences)


def count_letters(text):
    letter_count = 0
    for i in range(len(text)):
        if text[i].isalpha and not text[i] == " " and not is_punct(text[i]) and not text[i] == "," and not text[i] == "'":
            letter_count += 1

    return letter_count


def count_words(text):
    word_count = text.count(" ") + 1

    return word_count


def count_sentences(text):
    sentence_count = 0
    for i in range(len(text)):
        if is_punct(text[i]):
            sentence_count += 1

    return sentence_count


def calculate_reading_level(letters, words, sentences):
    L = letters / words * 100
    S = sentences / words * 100
    index = 0.0588 * L - 0.296 * S - 15.8

    index = round(index)

    if index > 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {index}")


def is_punct(character):
    if character == "." or character == "?" or character == "!":
        return True


main()